File: examples/Java/flserver2.java
package guide;

import org.zeromq.*;
import org.zeromq.ZMQ.Socket;

//  Freelance server - Model 2
//  Does some work, replies OK, with message sequencing
public class flserver2
{
    public static void main(String[] args)
    {
        if (args.length < 1) {
            System.out.printf("I: syntax: flserver2 <endpoint>\n");
            System.exit(0);
        }

        try (ZContext ctx = new ZContext()) {
            Socket server = ctx.createSocket(SocketType.REP);
            server.bind(args[0]);

            System.out.printf("I: echo service is ready at %s\n", args[0]);
            while (true) {
                ZMsg request = ZMsg.recvMsg(server);
                if (request == null)
                    break; //  Interrupted

                //  Fail nastily if run against wrong client
                assert (request.size() == 2);

                ZFrame identity = request.pop();
                request.destroy();

                ZMsg reply = new ZMsg();
                reply.add(identity);
                reply.add("OK");
                reply.send(server);
            }

            if (Thread.currentThread().isInterrupted())
                System.out.printf("W: interrupted\n");
        }
    }
}

File: examples/C++/flserver2.cpp
//  Freelance server - Model 2
//  Does some work, replies OK, with message sequencing

#include <iostream>
#include <zmqpp/zmqpp.hpp>

int main(int argc, char *argv[]) {
    if (argc < 2) {
        std::cout << "I: syntax: " << argv[0] << " <endpoint>" << std::endl;
        return 0;
    }
    zmqpp::context context;
    zmqpp::socket server(context, zmqpp::socket_type::reply);
    server.bind(argv[1]);

    std::cout << "I: echo service is ready at " << argv[1] << std::endl;
    while (true) {
        zmqpp::message request;
        try {
            server.receive(request);
        } catch (zmqpp::zmq_internal_exception &e) {
            if (e.zmq_error() == EINTR)
                std::cout << "W: interrupted" << std::endl;
            else
                std::cout << "E: error, errnum = " << e.zmq_error() << ", what = " << e.what()
                          << std::endl;
            break;  // Interrupted
        }
        //  Fail nastily if run against wrong client
        assert(request.parts() == 2);

        uint identity;
        request.get(identity, 0);
        // std::cout << "Received sequence: " << identity << std::endl;

        zmqpp::message reply;
        reply.push_back(identity);
        reply.push_back("OK");

        server.send(reply);
    }
    return 0;
}
